bleak/__init__.py

```diff
@@ -204,6 +205,33 @@ def set_scanning_filter(self, **kwargs):
         )
         self._backend.set_scanning_filter(**kwargs)
 
+    async def observe(self) -> AsyncIterator[Tuple[BLEDevice, AdvertisementData]]:
+        """
+        Yields devices and their advertising data as they are discovered.
+
+        .. note::
+            Ensure that scanning is started before calling this method.
+
+        Returns:
+            An async iterator that yields tuples (:class:`BLEDevice`, :class:`AdvertisementData`).
+        """
+        # noinspection PyProtectedMember
+        if self._backend._callback:
+            raise BleakError(
+                "Cannot use async iterator methods when detection_callback is used"
+            )
+
+        devices = asyncio.Queue()
+
+        self._backend.register_detection_callback(
+            lambda bd, ad: devices.put_nowait((bd, ad))
+        )
+        try:
+            while True:
+                yield await devices.get()
```

Comment:
We should add a `TODO` comment here that if/when https://github.com/hbldh/bleak/pull/1146 gets merged, we need to exit the generator here instead of waiting forever.

Reply:
And actually we probably need something now, to handle the case, e.g. if someone manually stops the scanner while we are still iterating.

Reply:
I will need to think about how to solve this neatly, if at all possible without adding some `stop` callbacks (`handle_early_stop`). Considering that the code will block on `async for _ in scanner.observe()`, the user cannot stop the scan from the same thread/coroutine, but of course it can be stopped in other ways, or as a result of external events. 

I wanted to add "How do we handle the unexpected stops now, if the user uses the callback?", but then I membered that this is what https://github.com/hbldh/bleak/pull/1146 is all about. Maybe we shall jump to that one first?

Reply:
I think for now we can leave this the way it is. After thinking about it some more, I'm not sure that stopping the generator when scanning stops is the most intuitive or desirable behavior. For example, I could see someone wanting to just set up this generator and leave it running forever in a long running app and connect the scanner start() and stop() methods to UI button events.

Instead, I'm thinking we might want a more general async context manager for capturing scanner stop events similar to how asyncio.timeout() works. This way users can selectively opt in to having a task interrupted when the scanner stops.